### DISCUSSION TITLE
Problem reading array from external .json file in a function

### DISCUSSION THREAD

**Discussion by ojfernandez** [nlohmann/json#3172] [Q&A] [answered]
I'm trying to read information from an external .json file passed as a parameter in a function (in a separate .cpp file), but it keeps giving me a 305 error (cannot use operator[] with a string argument in an array). I had this code in my main.cpp and it worked fine. I am also using another library called dpp but that shouldn't affect the json code.

**crows.json**
```
{
	"facts" : [
		{ "fact" : "Young crows leave the nest after four weeks." },
		{ "fact" : "Crows have around 250 different calls." },
		{ "fact" : "Crow eggs hatch after 20-40 days." },
		{ "fact" : "Crows are COOL!" },
		{ "fact" : "Crows are part of the family Corvidae." },
		{ "fact" : "Crows are omnivores." }
	],
	"imgs" : [
		{ "url" : "url.1" },
		{ "url" : "url.2" },
		{ "url" : "url.3" },
		{ "url" : "url.4" }
	]
}
```

**crowMsg.cpp**
```
#include <string>

#include "crowMsg.h"

using json = nlohmann::json;
using namespace std;

dpp::embed crowMsg(const json j, int& fact, int& factLast, int& img, int& imgLast) {
   /* Same number pull prevention */
   /* pulls a random number, tries again if number was pulled last */
   while (fact == factLast || img == imgLast) {
      fact = rand() % j.at("facts").size();
      img = rand() % j.at("imgs").size();
   }
   factLast = fact;
   imgLast = img;

   /* Gets fact and URL from .json */
   string crowFact = j["facts"][fact]["fact"];
   string crowImg = j["imgs"][img]["URL"];

   /* Creates an embed, shouldn't be affected */
   dpp::embed crowEmbed = dpp::embed().
   set_color(0x4b2e83).
   set_title("Crow Fact!").
   set_description(crowFact).
   set_image(crowImg).
   set_timestamp(time(0));
   
   return crowEmbed;
}
```

**crowMsg.h**
```
#ifndef CROWMSG_H
#define CROWMSG_H

#include <dpp/dpp.h>
#include <dpp/nlohmann/json.hpp>

dpp::embed crowMsg(const nlohmann::json, int&, int&, int&, int&);

#endif
```

**Comment by gregmarr** [comment, marked as answer]
Wouldn't cause this, but you're looking for `"URL"` when your JSON has `"url"`.  Also, you're passing `json` by value to the function, which can be very expensive, should pass by reference.

The most likely explanation is that `j` doesn't contain what you think it does.  Have you done something like this to verify?

    dpp::embed crowMsg(const json j, int& fact, int& factLast, int& img, int& imgLast) {
        std::cout << j.dump() << "\n";
 
You should also determine which line is causing the error, which will help narrow it down.

**Reply by nlohmann** [reply]
Also the extended diagnostics may help: https://json.nlohmann.me/home/exceptions/#extended-diagnostic-messages

**Reply by ojfernandez** [reply]
Thank you @gregmarr for suggesting .dump(). Turns out that it was reading a different file that I was calling in my main.cpp :)
